intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt

```diff
@@ -43,46 +46,82 @@ class HubPluginManager private constructor() : PluginManager<HubPlugin> {
+    if (!manifestFile.exists()) {
+      return PluginCreationFail(PluginDescriptorIsNotFound(DESCRIPTOR_NAME))
     }
-    return PluginCreationFail(PluginDescriptorIsNotFound(DESCRIPTOR_NAME))
+    val manifestContent = manifestFile.readText()
+    val manifest = Json(
+        JsonConfiguration.Stable.copy(isLenient = true, ignoreUnknownKeys = true)
+    ).parse(HubPluginManifest.serializer(), manifestContent)
+    val icon = try {
+      loadIcon(pluginDirectory, manifest.iconUrl)
+    } catch (e: IllegalArgumentException) {
+      return PluginCreationFail(HubIconInvalidUrl())
+    }
+    return createPlugin(manifest, manifestContent, icon)
   }
 
-  private fun loadDescriptor(descriptorFile: File): PluginCreationResult<HubPlugin> {
+  private fun createPlugin(
+      descriptor: HubPluginManifest, manifestContent: String, icon: PluginIcon?
+  ): PluginCreationResult<HubPlugin> {
     try {
-      val manifestContent = descriptorFile.readText()
-      val descriptor = Json(JsonConfiguration.Stable.copy(isLenient = true, ignoreUnknownKeys = true)).parse(HubPlugin.serializer(), manifestContent)
-      descriptor.manifestContent = manifestContent
-      val vendorInfo = parseHubVendorInfo(descriptor.author)
-      descriptor.apply {
-        vendor = vendorInfo.vendor
-        vendorEmail = vendorInfo.vendorEmail
-        vendorUrl = vendorInfo.vendorUrl
-      }
       val beanValidationResult = validateHubPluginBean(descriptor)
       if (beanValidationResult.any { it.level == PluginProblem.Level.ERROR }) {
         return PluginCreationFail(beanValidationResult)
       }
-      return PluginCreationSuccess(descriptor, beanValidationResult)
+
+      val vendorInfo = parseHubVendorInfo(descriptor.author ?: "")
+      if (vendorInfo.vendor == null) return PluginCreationFail(PropertyNotSpecified("author"))
+      val plugin = with(descriptor) {
+        HubPlugin(
+            pluginId = pluginId,
+            pluginName = pluginName,
+            pluginVersion = pluginVersion,
+            url = url,
+            description = description,
+            dependencies = dependencies,
+            products = products,
+            vendor = vendorInfo.vendor,
+            vendorEmail = vendorInfo.vendorEmail,
+            vendorUrl = vendorInfo.vendorUrl,
+            manifestContent = manifestContent,
+            icons = if (icon != null) listOf(icon) else emptyList()
+        )
+      }
+      return PluginCreationSuccess(plugin, beanValidationResult)
     } catch (e: Exception) {
       e.rethrowIfInterrupted()
       LOG.info("Unable to read plugin descriptor $DESCRIPTOR_NAME", e)
       return PluginCreationFail(UnableToReadDescriptor(DESCRIPTOR_NAME, e.localizedMessage))
     }
   }
+
+  private fun loadIcon(pluginDirectory: File, iconUrl: String?): PluginIcon? {
+    return when {
+      iconUrl == null -> null
+      iconUrl.contains("://") || iconUrl.startsWith("/") -> {
+        LOG.warn("Unsupported widget iconUrl: '$iconUrl'")
```

Comment:
If we throw an exception when we can not find the icon let's throw an exception here. and the icon becomes not null value

Reply:
It's logic from our application. 
If icon url is an external url than it's not an error, we just will not load it. But if icon url point to an element in archive and this element is missing this is an error

Reply:
okay

Reply:
why warning level then? either "info" or no logging at all